Use extensions from configuration for react native module resolution

apexskier opened the pull request:
<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. The two fields below are mandatory. -->

<!-- Please remember to update CHANGELOG.md in the root of the project if you have not done so. -->

**Summary**

Currently, react native files with extensions other than `.js` aren't imported when running jest in a react native environment. This is because of a hardcoded `.js` in the jest resolver. It should use the configuration's `moduleFileExtensions`.

**Test plan**

For a jest config of
```js
module.exports = {
    globals: {
        "ts-jest": {
            tsConfigFile: "tsconfig.test.json",
        },
    },

    preset: "react-native",

    moduleFileExtensions: [
        "js",
        "jsx",
        "ts",
        "tsx",
    ],
};
```

a test of
```js
import testEnvironment from "../__mocks__/testImport";

describe("Native test environment", () => {
    it("finds the native version of the dependency", () => {
        expect(testEnvironment).toBe("native");
    });
});
```

`__mocks__/testImport`
```ts
export default "web";
```
`__mocks__/testImport.native`
```ts
export default "native";
```

Without this PR (on master)
```
$ jest ./src/__tests__/nativeTestEnvironment.js 
 FAIL  src/__tests__/nativeTestEnvironment.js
  Native test environment
    ✕ finds the native version of the dependency (7ms)

  ● Native test environment › finds the native version of the dependency

    expect(received).toBe(expected)
    
    Expected value to be (using ===):
      "native"
    Received:
      "web"
      
      at Object.<anonymous> (src/__tests__/nativeTestEnvironment.js:9:38)
      at tryCallTwo (node_modules/promise/lib/core.js:45:5)
      at doResolve (node_modules/promise/lib/core.js:200:13)
      at new Promise (node_modules/promise/lib/core.js:66:3)
      at tryCallOne (node_modules/promise/lib/core.js:37:12)
      at node_modules/promise/lib/core.js:123:15

Test Suites: 1 failed, 1 total
Tests:       1 failed, 1 total
Snapshots:   0 total
Time:        1.972s, estimated 2s
Ran all test suites matching /.\/src\/__tests__\/nativeTestEnvironment.js/i.
```

With this PR's code

```
$ jest ./src/__tests__/nativeTestEnvironment.js 
 PASS  src/__tests__/nativeTestEnvironment.js
  Native test environment
    ✓ finds the native version of the dependency (3ms)

Test Suites: 1 passed, 1 total
Tests:       1 passed, 1 total
Snapshots:   0 total
Time:        1.445s, estimated 2s
Ran all test suites matching /.\/src\/__tests__\/nativeTestEnvironment.js/i.
```

cpojer commented on packages/jest-resolve/src/index.js:
```diff
@@ -117,13 +117,19 @@ class Resolver {
-      extensions.unshift('.' + NATIVE_PLATFORM + '.js');
-    }
+
+    let extensions: Array<string> = [];
```
const. You don't need the `Array<string>`, flow will be able to figure it out.

cpojer commented on packages/jest-resolve/src/index.js:
```diff
@@ -117,13 +117,19 @@ class Resolver {
+        this._options.extensions.map(ext => '.' + NATIVE_PLATFORM + ext),
+      );
     }
+    extensions = extensions.concat(this._options.extensions);
```
`extensions.push(...this._options.extensions)`

cpojer commented:
Please add a test like @SimenB and @thymikee requested. Also, can you point me to the place in https://github.com/facebook/metro-bundler where the behavior matches what you are doing in this PR? Also cc @rafeca since we talked about this yesterday.

apexskier commented:
Added tests and updated code style, thanks for the feedback.

@cpojer I haven't looked at the source of metro-bundler before, but I think the matching behavior is in the dependency graph module resolution at https://github.com/facebook/metro-bundler/blob/c6b0134398bdd4ded4a7efbb08e44bcd4bf10b7a/packages/metro-bundler/src/node-haste/DependencyGraph/ModuleResolution.js#L440-L465

cpojer commented:
Thanks for making the changes to Jest. We are currently looking into plucking jest-resolve into Metro so we can get rid of the resolution logic there. Aligning the two is critical to make this work.
